## Numpy assignment

## 1 Preliminaries

In [201]:
# The usual preamble
import numpy as np

## Numpy

In [3]:
import numpy as np
x = np.array([2,3,1,0])
y =np.arange(10)
z = 2 + x

Write expressions in the next three cells to retrieve 1 from `x`, 3 from `y`, and 3 from `z`.

In [4]:
x[2]

1

In [6]:
y # See the content
y[3]

3

In [9]:
z # See the content
z[2]

3

In the next cell, write an expression that generates a 5 by 5 array filled with zeros, and another  that generates a 1 by 1 array with a 1 in it.

In [15]:
M_5x5 = np.zeros((5, 5))
M_5x5

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [18]:
M_1x1 = np.ones((1, 1))
M_1x1

array([[1.]])

In the next cell write an expression whose value is a 1D array of length 25 filled with the even integers from 2 to 50 inclusive and reshape it into a 5x5 array.  Assign this array to the name `X`; `X` should evaluate as follows.

```
In [ ]: X

Out [ ]: array([[ 2,  4,  6,  8, 10],
                [12, 14, 16, 18, 20],
                [22, 24, 26, 28, 30],
                [32, 34, 36, 38, 40],
                [42, 44, 46, 48, 50]])
```


Write an expression that retrieves the following subarray of `X`.  Use slicing in both dimensions
of `X`.

```
array([[24, 26],
       [34, 36]])
```

In [20]:
X = np.arange(2,52,2).reshape((5,5))
X

array([[ 2,  4,  6,  8, 10],
       [12, 14, 16, 18, 20],
       [22, 24, 26, 28, 30],
       [32, 34, 36, 38, 40],
       [42, 44, 46, 48, 50]])

In [28]:
X[2:4, 1:3]

array([[24, 26],
       [34, 36]])

Retrieve the following from `X`

```
array([[18, 20],
       [38, 40]])
```

Hint: Use a slice with steps.

In [37]:
X[1::2, 3:5]

array([[18, 20],
       [38, 40]])

Use "fancy indexing" to write one expression that retrieves the following values from X

```
array([12, 26, 50])
```

In [38]:
X[[1, 2, -1], [0, 2, -1]]

array([12, 26, 50])

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  

#### Exercise

The next cell defines two arrays `a` and `b`.
In this exercise , your task is to create an array
that is exactly like `a` except that the negative values in `a` are replaced by
`0`. And similarly for `b`. 

You can do this either with assignments (which
will change `a` and `b`) or with a single expression that will create 
new arrays with the negative values replaced by `0`s.

The challenge is this: Make sure that the whatever solution you come up
for `a` also works for `b`, with `b` substituted for `a.

Here are `a` and `b`

In [98]:
a = np.arange(10,-1,-1) - 3
print(a)
b = np.array([1,-1] * 5) * np.arange(1,11)
print(b)

[ 7  6  5  4  3  2  1  0 -1 -2 -3]
[  1  -2   3  -4   5  -6   7  -8   9 -10]


One assignment that solves the problem for `a` is:

In [ ]:
a[-3:] = 0
a

array([7, 6, 5, 4, 3, 2, 1, 0, 0, 0, 0])

This replaces all negative values in `a` with `0`s, as desired.

But this isn't a good overall solution, because this
assignment doesn't eliminate the negative values in  `b`.

In [ ]:
b[-3:] = 0
b

array([ 1, -2,  3, -4,  5, -6,  7,  0,  0,  0])

There are many negative values left.

See if you can cook up a solution that applies
to both `a` and `b` and gives us arrays with `0` replacing all
the negative values.  

**Hint One**: Use a Boolean condition
combined with an assignment.  This is the strategy that
changes both `a` and `b`.

In [99]:
# Your answer here if you used Hint One
print("a: ", a) # See the content
print("b: ", b) # See the content
a[a < 0] = 0
print("Non-negative a:", a)
b[b < 0] = 0
print("Non-negative b:", b)

a:  [ 7  6  5  4  3  2  1  0 -1 -2 -3]
b:  [  1  -2   3  -4   5  -6   7  -8   9 -10]
Non-negative a: [7 6 5 4 3 2 1 0 0 0 0]
Non-negative b: [1 0 3 0 5 0 7 0 9 0]


**Hint Two**:  One option that doesn't change `a` and `b` is to use `np.where`.


The form for using `np.where` is:

```
>>> np.where(condition, x, y)
```

where `condition`, `x` and `y` are arrays of the same size (or are broadcastable to the same size).
Let `a = np.where(condition, x, y)`;
then `a[i] == x[i]` if `condition[i] == True`; otherwise `a[i] ==y[i]`.

For example
```
>>> np.where(np.array([True,False,True]),np.array([1,23,3]),np.array([0,2,0]))
array([1, 2, 3])
```

Another example:  Given `a` and `b` as defined below,  return an array just like `a` wherever
`a[i]` is the square root of `b[i]` and `0` where `a[i]` is not.

```
>>> a = np.array([1, 4.1, 9, 16])
>>> b = np.array([1,2,3,4])
>>> condition = (np.sqrt(a) == b)
>>> np.where(condition,a,0)
array([1., 0, 9., 16.])
```

Note that `condition` is just a Boolean array as in the previous example:

```
>>> condition
array([ True, False,  True,  True])
```

In [100]:
# Your answer here if you used Hint Two
print("a: ", a) # See the content
print("b: ", b) # See the content
conditiona = (a >= 0)
a_pos = np.where(conditiona, a, 0)
print("Non-negative a:",  a_pos)
conditionb = (b >= 0)
b_pos = np.where(conditionb, b, 0)
print( "Non-negative b:",  b_pos)

a:  [7 6 5 4 3 2 1 0 0 0 0]
b:  [1 0 3 0 5 0 7 0 9 0]
Non-negative a: [7 6 5 4 3 2 1 0 0 0 0]
Non-negative b: [1 0 3 0 5 0 7 0 9 0]


#### Exercise

In the next cell write an expression that produces an array containing result of multiplying 7 by each of the first 12 integers (1 - 12). There's a hard way to do this and an easy way. The hard way uses a loop. The easy way uses `np.arange` and elementwise multiplication. Don't use the hard way.

In [109]:
arr = np.arange(1, 13) # Array of 12 integers: 1 - 12
print("Array of integers 1 - 12 before multiplication: ", arr)
arr_x7 = arr * 7
print("Array of integers 1 - 12 after multiplication: ", arr_x7)

Array of integers 1 - 12 before multiplication:  [ 1  2  3  4  5  6  7  8  9 10 11 12]
Array of integers 1 - 12 after multiplication:  [ 7 14 21 28 35 42 49 56 63 70 77 84]


## 2D arrays

In [111]:
import numpy as np
room_matrix = \
np.array(
[[6,  1, 4, 2],
[5,  2, 2, 2],
[7,  3, 3, 2],
[5,  5, 3, 1],
[8, 7, 5, 2]])

cost_vector = np.array([40, 175, 90, 450])

In the next cell write an express that retrieves the last column of `room_matrix`.

In [120]:
room_matrix[:,-1]

array([2, 2, 2, 1, 2])

In the next cell, write an expression that retrieves the third row of `room_matrix`.

In [121]:
room_matrix[2,:]

array([7, 3, 3, 2])

In the next cell write an expression that retrieves the first two elements of the second column of `room_matrix`

In [150]:
room_matrix[:2,1]

array([1, 2])

In the next cell write an expression that retrieves the following 2D subarray from `room_matrix`:

```
array([[2,2],
       [3,2]])
```

In [225]:
print(room_matrix) # Display the content for clarity
room_matrix[1:3, 1::2]

[[6 1 4 2]
 [5 2 2 2]
 [7 3 3 2]
 [5 5 3 1]
 [8 7 5 2]]


array([[2, 2],
       [3, 2]])

### In the next cell, write an expression that returns a Boolean array that identifies the members of `room_matrix` which are greater than or equal to 7.

In [230]:
room_matrix # Display the content for clarity

array([[6, 1, 4, 2],
       [5, 2, 2, 2],
       [7, 3, 3, 2],
       [5, 5, 3, 1],
       [8, 7, 5, 2]])

In [231]:
room_matrix >= 7

array([[False, False, False, False],
       [False, False, False, False],
       [ True, False, False, False],
       [False, False, False, False],
       [ True,  True, False, False]])

In the next cell, write an expression that produces an array containing the items in `room_matrix` which are greater than or equal to 7.  Note:  This question and the previous question do not have the same answer.

In [232]:
room_matrix[room_matrix >= 7]

array([7, 8, 7])

## Iris data exercise

In the next cell, use a Boolean mask to return all the rows of the iris data set whose column 1 value is greater than 3.0 (I mean the column whose **index** is 1).
The next cell is not empty because you have been provided with the code which loads the data.
So you should keep that code and **add to it**.  Be sure to save the rows found by assigning
them to a variable name

In [233]:
from sklearn.datasets import load_iris
data = load_iris()
features = data['data']
target = data['target']

In [234]:
features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In the next cell, use a Boolean array to return all the rows representing members of class 2 in the iris data set (the Boolean array is not the answer; it is used in retrieving the answer).

In [236]:
target

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

In [235]:
target == 2

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,

In [242]:
target[target == 2]

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2])

In [263]:
features[target == 2] # I believe this is what was required - they way I understood: all the data corresponding to class 2 

array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.3, 2.9, 5.6, 1.8],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [4.9, 2.5, 4.5, 1.7],
       [7.3, 2.9, 6.3, 1.8],
       [6.7, 2.5, 5.8, 1.8],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [6.5, 3. , 5.5, 1.8],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6. , 2.2, 5. , 1.5],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.3, 2.7, 4.9, 1.8],
       [6.7, 3.3, 5.7, 2.1],
       [7.2, 3.2, 6. , 1.8],
       [6.2, 2.8, 4.8, 1.8],
       [6.1, 3. , 4.9, 1.8],
       [6.4, 2.8, 5.6, 2.1],
       [7.2, 3. , 5.8, 1.6],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [6.3, 2.8, 5.1, 1.5],
       [6.1, 2

**Extra credit**:  In the next cell write an expression -- or expressions -- which returns
all the rows of `features` representing members of class 2 whose column 3 value is greater than 1.8.

Two solutions: 

1. One way is to set a variable to the result of one of your answers to the last two questions. That result is an array, and you can do more indexing on it.  
2. Another way that expands yoiur skill set more is to use a Boolean operator to combine conditions (note you will need to wrap parentheses around your conditions before combining them with the Boolean operator).

In [266]:
# Attempt
arr = features[target == 2] # Storing the answer to the previous question in a new variable
col3 = arr[:,2] # Column 3
# Here I found that all the numbers in the 3rd column are greater than 1.8; maybe I messed up, but I checked several times and still looks good
# print(col3)
# print(col3[col3 > 1.8]) 
# To still do some filtering, I used the last column:
col4 = arr[:,3]
print(col4)
print(col4[col4 > 1.8]) 

[2.5 1.9 2.1 1.8 2.2 2.1 1.7 1.8 1.8 2.5 2.  1.9 2.1 2.  2.4 2.3 1.8 2.2
 2.3 1.5 2.3 2.  2.  1.8 2.1 1.8 1.8 1.8 2.1 1.6 1.9 2.  2.2 1.5 1.4 2.3
 2.4 1.8 1.8 2.1 2.4 2.3 1.9 2.3 2.5 2.3 1.9 2.  2.3 1.8]
[2.5 1.9 2.1 2.2 2.1 2.5 2.  1.9 2.1 2.  2.4 2.3 2.2 2.3 2.3 2.  2.  2.1
 2.1 1.9 2.  2.2 2.3 2.4 2.1 2.4 2.3 1.9 2.3 2.5 2.3 1.9 2.  2.3]


Help with Solution 2.  The following example illustrates wrapping parentheses around conditions.
Say you have X an array of numbers and you want a Boolean array
that tells you exactly which numbers are even (`X%2 == 0`) but not greater than 4 (`not (X > 4)`).  You can
combine these two conditions as follows:

In [ ]:
X = np.arange(-3,12).reshape((3,5))
print(X)
X[(X%2 == 0) & ~(X > 4)]

[[-3 -2 -1  0  1]
 [ 2  3  4  5  6]
 [ 7  8  9 10 11]]


array([-2,  0,  2,  4])

Your solution or solutions to the Extra Credit problem in the next cell.

In [350]:
# One more attempt
# arr = ((features[:,3][target == 2]) & (features[:,3] > 1.8)) # Not working
# print((features[:,3]) or (features[target == 2]))
features[(target == 2) & (features[:,3] > 1.8)] # Outputs all the rows where value in certain column is greater than 1.8
# arr = (features[:,3][target == 2]) & (features[:,3] > 1.8)


array([[6.3, 3.3, 6. , 2.5],
       [5.8, 2.7, 5.1, 1.9],
       [7.1, 3. , 5.9, 2.1],
       [6.5, 3. , 5.8, 2.2],
       [7.6, 3. , 6.6, 2.1],
       [7.2, 3.6, 6.1, 2.5],
       [6.5, 3.2, 5.1, 2. ],
       [6.4, 2.7, 5.3, 1.9],
       [6.8, 3. , 5.5, 2.1],
       [5.7, 2.5, 5. , 2. ],
       [5.8, 2.8, 5.1, 2.4],
       [6.4, 3.2, 5.3, 2.3],
       [7.7, 3.8, 6.7, 2.2],
       [7.7, 2.6, 6.9, 2.3],
       [6.9, 3.2, 5.7, 2.3],
       [5.6, 2.8, 4.9, 2. ],
       [7.7, 2.8, 6.7, 2. ],
       [6.7, 3.3, 5.7, 2.1],
       [6.4, 2.8, 5.6, 2.1],
       [7.4, 2.8, 6.1, 1.9],
       [7.9, 3.8, 6.4, 2. ],
       [6.4, 2.8, 5.6, 2.2],
       [7.7, 3. , 6.1, 2.3],
       [6.3, 3.4, 5.6, 2.4],
       [6.9, 3.1, 5.4, 2.1],
       [6.7, 3.1, 5.6, 2.4],
       [6.9, 3.1, 5.1, 2.3],
       [5.8, 2.7, 5.1, 1.9],
       [6.8, 3.2, 5.9, 2.3],
       [6.7, 3.3, 5.7, 2.5],
       [6.7, 3. , 5.2, 2.3],
       [6.3, 2.5, 5. , 1.9],
       [6.5, 3. , 5.2, 2. ],
       [6.2, 3.4, 5.4, 2.3]])

## Part One

### California Housing data exercise

In [299]:
from sklearn.datasets import fetch_california_housing
housing = fetch_california_housing()
X,Y = housing.data,housing.target
print(X.shape,Y.shape)

(20640, 8) (20640,)


The cell above loads the `sklearn` California Housing Dataset

### California Housing Data Set Characteristics

    :Number of Instances: 20640

    :Number of Attributes: 8 numeric, predictive attributes and the target

    :Attribute Information:
        - MedInc        median income in block group
        - HouseAge      median house age in block group
        - AveRooms      average number of rooms per household
        - AveBedrms     average number of bedrooms per household
        - Population    block group population
        - AveOccup      average number of household members
        - Latitude      block group latitude
        - Longitude     block group longitude

    :Missing Attribute Values: None

This dataset was obtained from the StatLib repository.
https://www.dcc.fc.up.pt/~ltorgo/Regression/cal_housing.html

The target variable is the median house value for California districts,
expressed in hundreds of thousands of dollars ($100,000).

This dataset was derived from the 1990 U.S. census, using one row per census
block group. A block group is the smallest geographical unit for which the U.S.
Census Bureau publishes sample data (a block group typically has a population
of 600 to 3,000 people).

An household is a group of people residing within a home. Since the average
number of rooms and bedrooms in this dataset are provided per household, these
columns may take surprisingly large values for block groups with few households
and many empty houses, such as vacation resorts.

.. topic:: References

    - Pace, R. Kelley and Ronald Barry, Sparse Spatial Autoregressions,
      Statistics and Probability Letters, 33 (1997) 291-297

The array `X`, defined in the cell above, is a 20,640x8 array containing the values of all
these variables for 20,640 California block groups.

The information stored in Y, the value to be predicted is

```
       - MEDV     Median value of owner-occupied homes in $1000's
 
```

#### Exercise

1.  Construct an expression or expressions that creates a new table that omits all blocks whose average number of bedrooms is greater than 10.
2.  Construct an expression or expressions that creates a new table containing all columns except the 
    `MedInc` column.  This will be our new X.  Also create a new Y containing the `MedInc` column.
    This is to help someone research how to predict median income using the other variables.  
    The shape of the new X should
    20,640x7 and the new Y should be a 1D array with length 20,640.  Show that you checked this in your answer.

In [41]:
# 1. Create a new table that omits all blocks whose average number of bedrooms is greater than 10
# AveBedrms - Column = 4, X[:,3] 
# X_omitted = X[X[:,3] > 10] # The data we want to omit
# print(X_omitted.shape)
X_updated = X[~(X[:,3] > 10)]
print(X_updated.shape, X.shape)

(20631, 8) (20640, 8)


In [275]:
# 2. Create a new table containing all columns except the MedInc column
# This will be our new X
# MedInc - Column = 1, X[:,0]

# X_new = X[:, ~np.all(X < 1, axis = 0)]
# print(X_new.shape)
# print(X_new)
# print(X)
# X_new = ((X) & (~X_tbomitted))


# X_tbomitted =X[:,0].astype(int)
# X_new =X[ ~(X_tbomitted)]
# X_new = X_new[np.logical_and(X[:,], (np.logical_not(X[:,0])))]

############ Finaly after so much struggle
X_new = X[:, 1:]
# print(X_new)
print("********************************************************************")
print("X's 1st column content:", X[:,0])
print("X_new's shape:", X_new.shape)
print("X_new's content:", X_new)
print("********************************************************************")
# Create a new Y containing the MedInc column
Y_new = X[:,0]
print("Y_new's shape:", Y_new.shape)
print("Y_new's content", Y_new)

********************************************************************
X's 1st column content: [8.3252 8.3014 7.2574 ... 1.7    1.8672 2.3886]
X_new's shape: (20640, 7)
X_new's content: [[  41.            6.98412698    1.02380952 ...    2.55555556
    37.88       -122.23      ]
 [  21.            6.23813708    0.97188049 ...    2.10984183
    37.86       -122.22      ]
 [  52.            8.28813559    1.07344633 ...    2.80225989
    37.85       -122.24      ]
 ...
 [  17.            5.20554273    1.12009238 ...    2.3256351
    39.43       -121.22      ]
 [  18.            5.32951289    1.17191977 ...    2.12320917
    39.43       -121.32      ]
 [  16.            5.25471698    1.16226415 ...    2.61698113
    39.37       -121.24      ]]
********************************************************************
Y_new's shape: (20640,)
Y_new's content [8.3252 8.3014 7.2574 ... 1.7    1.8672 2.3886]


#### Exercise

Write an expression that computes how many blocks in the data set 
have an average house age greater than 25 years old and have a population greater than 5,000.

Note there are two solutions.  One involves finding the set of rows that
satisfy the constraint and computing its length.  Another 
finds the Boolean mask that represents those rows and computes
the number of `True`s in it.  See if you can find both

In [358]:
# This implementation should find the Boolean mask that represents those rows and computes the number of Trues in it
HouseAge = X[:,1] > 25
# print(HouseAge.sum())
# print(HouseAge.shape)
Population = X[:,4] > 5000
# print(Population.sum())
# print(Population.shape)
find = (HouseAge) & (Population)
print(find.sum())

23


In [366]:
# This implementation should  involve finding the set of rows that satisfy the constraint and computing its length
HouseAge = X[:,1] > 25
# print(HouseAge.sum())
# print(HouseAge.shape)
Population = X[:,4] > 5000
# print(Population.sum())
# print(Population.shape)
find = X[(HouseAge & Population)]
print(find.shape[0])
print(len(find))

23
23


Checks!

## Part two

The answers for the following section are available on the inclass Boolean exercises solution notebook.

# Boolean combinations

In this section we load forest covertype data.  This is a multi class dataset that has data for 7 different forest covertypes (stored in the `target` attribute).  There are 581,012 forest plots with 54 attributes each (stored ina 581012x54 array) The first ten attributes are numerical, the last 44 are Boolean (true/false) attributes. Each of the Boolean attributes represents a qualitative soil type attribute which is either present or absent.  We will refer to all the attributes by their column index.  Thus the first attribute is attribute 0 and the last (a Boolean attribute) is attribute 53.

In [ ]:
#from sklearn.datasets import load_wine
#wdata = load_wine()
from sklearn.datasets import fetch_covtype
data = fetch_covtype()
print(data.data.shape)  # data.data is the 581012 x 54 array
print(data.target.shape)     # data.target contains the class for each instance

Th first 26 columns of the plot index 50_210:

In [2]:
data.data[50_210,:26]

array([2.988e+03, 2.250e+02, 2.000e+00, 5.520e+02, 9.200e+01, 5.492e+03,
       2.160e+02, 2.410e+02, 1.620e+02, 4.290e+03, 1.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00, 0.000e+00,
       0.000e+00, 0.000e+00])

The last 28 column of the plot index 50_210:

In [3]:
data.data[50_210,26:]

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 1.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])

[The Boolean arrays and masks notebook](https://github.com/gawron/python-for-social-science/blob/master/numpy/02_06_Boolean_Arrays_and_Masks.ipynb) discusses combining Boolean arrays with Boolean operators `&` (conceptually 'and') and `|` (conceptually 'or').   We can add to that `~` (conceptually 'not') to complete the list.  Look at the examples there.

 
The `sklearn` description of the covertype dataset is printed in the cell below. For a fuller understanding of the attributes in the dataset you might want to look at [the original UCI data set description.](https://archive.ics.uci.edu/ml/datasets/Covertype)

Forest covertypes
---------------------

The samples in this dataset correspond to 30×30m patches of forest in the US,
collected for the task of predicting each patch's cover type,
i.e. the dominant species of tree.
There are seven covertypes, making this a multiclass classification problem.
Each sample has 54 features, described on the
[dataset's homepage](https://archive.ics.uci.edu/ml/datasets/Covertype>).
Some of the features are boolean indicators
while others are discrete or continuous measurements.

**Data Set Characteristics:**

    =================   ============
    Classes                        7
    Samples total             581012
    Dimensionality                54
    Features                     int
    =================   ============

The function `sklearn.datasets.fetch_covtype` will load the covertype dataset;
it returns a dictionary-like object
with the feature matrix in the ``data`` member
and the target values in ``target``.
The dataset will be downloaded from the web if necessary.

`data.data` attribute is a `numpy` array containing the array.  As an example to help with this problem.  To compute an an array containing just the rows in which attribute 0 is greater than 50 you do. 

In [4]:
WW = data.data > 50
print(WW.shape)

(581012, 54)


The covertypes (or dominant tree) for each forest plot are in `data.target`, which is often the attribute used to store the classes in an `sklearn` clasification dataset.  There is data for 581,012 plots
so this is a 1D array of length 581,012.  

There are 7 distinct fiorest tyoes that a plot can have.

In [5]:
print(len(data.target))
print(set(data.target))

581012
{1, 2, 3, 4, 5, 6, 7}


### The problem

1. Write an expression that retrieves all rows that a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

2. Write an expression that retrieves the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

3. Write an expression that counts the rows from either class 3 or class 5  that have a value greater than 300 in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12. 

In [30]:
# Write an expression that retrieves all rows that a value greater than 300 in the column indexed 1 (second column) and 
# a value equal to 1 for the column indexed 12
myData01 = data.data[:,1] > 300
print(myData01.sum())
# print(data.data[myData01])
myData02 = data.data[:,12] == 1.
print(myData02.sum())
# print(data.data[myData02])
myData0 = (myData01) & (myData02)
print(myData0.sum())
# print(data.data[myData02])

102696
253364
45076


In [43]:
# Write an expression that retrieves the rows from either class 3 or class 5 that have a value greater than 300 
# in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12
myData10 = ~((data.target==3) | (data.target==5))
# print(myData03.sum())
# print(data.data[myData03])
myData1 = (myData10) & (myData0)
print(data.data[myData1])

[[2780.  346.   13. ...    0.    0.    0.]
 [2725.  353.   19. ...    0.    0.    0.]
 [2747.  329.   19. ...    0.    0.    0.]
 ...
 [2619.  336.   13. ...    0.    0.    0.]
 [2630.  317.   12. ...    0.    0.    0.]
 [2633.  309.    9. ...    0.    0.    0.]]


In [44]:
# Write an expression that counts the rows from either class 3 or class 5 that have a value greater than 300 
# in the column indexed 1 (second column) and a value equal to 1 for the column indexed 12
print(myData1.sum())

43815
